Dockerfile

```diff
@@ -1,19 +1,85 @@
-#--------- Generic stuff all our Dockerfiles should start with so we get caching ------------
+##############################################################################
```

Comment:
RFC to @NyakudyaA .

We are going to try to leverage Dockerfile Multi-stage build. I don't see any reasoning not too. It's more convenient, easier to maintain, and also cacheable. The only drawback that I regret is they can't be separate with some kind of `include` mechanism.

TL;DR; on how to leverage multi stage build:

We combine previous 3 Dockerfile (base, prod, and testing) into one but with 3 stages (base, prod, and testing).

Append extra parameters `--target` into the docker build (or docker-compose build) to specify which target you want to build.

Since we can refer the image inside the recipe (for example `postgis-base` is the name of the base image stage), caching happens without us having to push the image into registry. For example the  prod image is built using `FROM postgis-base AS postgis-prod`. We can do build like this:

```
# Build base image first (like we usually do)
docker build --target postgis-base .
# Build prod image, but it will start by using the cache from the postgis-base stage (no need to rebuild base)
docker build --target postgis-prod .
```

Or like this:
```
# Build prod image directly
docker build --target postgis-prod .
# Then build the base image as a tagged image, because we want to push the base over to registry
docker build -t kartoza/postgis:base --target postgis-base .
# In the above command, docker will use build cache because it already build postgis-base stage, when it builds postgis-prod
```

Basically we can keep the same usage workflow as usual, but with caching improvement in local and cloud build. The only thing we need to change is the build scripts.